.github/workflows/ci.yaml

```diff
@@ -1,10 +1,6 @@
 name: CI
 
-on:
-  push:
-    branches: [main]
-  pull_request:
-    branches: [main]
+on: [push, pull_request, workflow_dispatch]
```

Comment:
The issue with this configuration is that it always results in duplicate CI runs for pull requests: once [for the `push` event](https://github.com/hugovk/ruff/actions/runs/4007012752) in the base repository and once [for the `pull_request` event](https://github.com/charliermarsh/ruff/actions/runs/4007137037) in the head repository.

Reply:
Yes, I think that's fine. 

First, so contributors can see test results it on their forks first, to make sure it passes before proceeding; and then again for the base repo, so the test results can be seen by maintainers in the PR.

Reply:
Hmm, yeah, this actually does seem non-ideal, since PRs in the main repo are now showing all the duplicate tasks. It'd be nice to at least get rid of those somehow.

Reply:
I’d recommend

```yaml
on:
  push:
    branches: [main]
  pull_request:
  workflow_dispatch:
```

and in the unusual case where a contributor wants to run CI on their fork before creating a pull request, `workflow_dispatch` allows them to [trigger it manually](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow#running-a-workflow).

Reply:
Thank you.

Reply:
Hi!

`workflow_dispatch` is good for manual triggers, but as a contributor who wants to test my code on CI before opening PRs, the amount of clicks it requires to start each build is pretty tedious. And CI should be automatic not manual.

CIs are amazing things - you can automatically test and lint your code on all these different versions! We should encourage contributors to use them. But if they don't want to, that's fine, and GitHub Actions is normally disabled for forks.

As another approach, how does this config look?

```yml
    # We want to run on external PRs, but not on our own internal PRs as they'll be run
    # by the push to the branch.
    if: github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository
```

re: https://github.com/has2k1/plotnine/blob/f696c01f1602e422a29778c112005463992ed707/.github/workflows/testing.yml#L10-L12

Reply:
I don't really see the problem with having to open a PR to run the CI ... you can open the PR as a draft and mark it as ready once the CI has passed if you care about that.

Reply:
One reason is to avoid noise on the tracker, for example, there may be lots of commits/rebases etc as things progress.

Another option is to follow the lead of things like pytest and pre-commit, and run the CI for branches prefixed with `test-me-`:

```yml
on:
  push:
    branches: [main, test-me-*]
```

https://github.com/pytest-dev/pytest/blob/ca40380e99c2cdaab1d0c041f9f28cff37ef8ff9/.github/workflows/test.yml#L3-L16

https://github.com/pre-commit/pre-commit/blob/dd8e717ed6022209a2b0cecf5c75460eb60e548e/.github/workflows/main.yml#L3-L5

Reply:
You could open a PR in your own fork.